### ISSUE TITLE
[BUG] Constructing a ProtoType without a field value makes the object behave as if it had not this property

### ISSUE THREAD

**Issue by PhilipVinc** [JuliaIO/ProtoBuf.jl#146]
I'm moving [tensorboardlogger](https://github.com/PhilipVinc/TensorBoardLogger.jl/pull/80) over to ProtoBuf 0.9 and regenerating all protobuf objects.

Among several proto types there is Event defined as follows in [event.proto](https://github.com/tensorflow/tensorboard/blob/master/tensorboard/compat/proto/event.proto)
```
syntax = "proto3";

package tensorboard;

import "tensorboard/compat/proto/summary.proto";

// Protocol buffer representing an event that happened during
// the execution of a Brain model.
message Event {
  // Timestamp of the event.
  double wall_time = 1;

  // Global step of the event.
  int64 step = 2;

  oneof what {
    // An event file was started, with the specified version.
    // This is use to identify the contents of the record IO files
    // easily.  Current version is "brain.Event:2".  All versions
    // start with "brain.Event:".
    string file_version = 3;
    // An encoded version of a GraphDef.
    bytes graph_def = 4;
    // A summary was generated.
    Summary summary = 5;
    // The user output a log message. Not all messages are logged, only ones
    // generated via the Python tensorboard_logging module.
    LogMessage log_message = 6;
    // The state of the session which can be used for restarting after crashes.
    SessionLog session_log = 7;
    // The metadata returned by running a session.run() call.
    TaggedRunMetadata tagged_run_metadata = 8;
    // An encoded version of a MetaGraphDef.
    bytes meta_graph_def = 9;
  }
}
```

I then generated with Protobuf.jl@0.9 the proto.jl files, (you can test with TensorBoardLogger#threadsafe).
However, if I try to access a field I did not provide to the constructor, they are not initialised to the default value but they are not present!
```julia
# ]add TensorBoardLogger#threadsafe 

using TensorBoardLogger
TensorBoardLogger.tensorboard.Event()

julia> ev.step
ERROR: KeyError: key :step not found
Stacktrace:
 [1] getindex(::Dict{Symbol,Any}, ::Symbol) at ./dict.jl:467
 [2] getproperty(::TensorBoardLogger.tensorboard.Event, ::Symbol) at /Users/filippovicentini/Dropbox/Ricerca/Codes/Julia/TensorBoardLogger/src/protojl/TensorBoard/event_pb.jl:185
 [3] top-level scope at REPL[6]:1
```

However, `ev.step` should be initialised to 0 according to google's documentation, as it is not an optional field

**Comment by PhilipVinc** [comment]
Sorry for the bump @tanmaykm, could you please let me know if this is fixable or point me in the right direction to fix it?
It's the only thing holding up a new release of TensorBoardLogger

**Comment by tanmaykm** [comment]
Hi @PhilipVinc, sorry had missed this earlier.

Default properties are set on an object that is parsed from a message as stated [here](https://developers.google.com/protocol-buffers/docs/proto3#default). They are not set by default when one is constructed directly in Julia code. They can however be set in your code simply by calling the `setdefaultproperties!` method:

```julia
julia> ev = ProtoBuf.setdefaultproperties!(Event());

julia> ev.step
0
```

**Comment by tanmaykm** [comment]
Yes, that shouldn't be happening because `readproto` does invoke `setdefaultproperties!` [here](https://github.com/JuliaIO/ProtoBuf.jl/blob/1378c79f04750971506e67dd380ebddd0077101e/src/codec.jl#L536). Strange.

**Comment by PhilipVinc** [comment]
Ah, indeed.

```
julia> ev=TensorBoardLogger.tensorboard.Event();
julia> ProtoBuf.setdefaultproperties!(ev);
julia> ev.step
ERROR: KeyError: key :step not found
Stacktrace:
 [1] getindex(::Dict{Symbol,Any}, ::Symbol) at ./dict.jl:467
 [2] getproperty(::TensorBoardLogger.tensorboard.Event, ::Symbol) at /Users/filippovicentini/Dropbox/Ricerca/Codes/Julia/TensorBoardLogger/src/protojl/tensorboard/event_pb.jl:185
 [3] top-level scope at REPL[15]:1
```

**Comment by PhilipVinc** [comment]
I don't know if it's relevant, but I can't find `step` in Event's meta structure:

```julia
julia> meta(TensorBoardLogger.Event)
ProtoMeta for TensorBoardLogger.tensorboard.Event
ProtoMetaAttribs[ProtoMetaAttribs(1, :wall_time, :double, Float64, 0, false, [0.0], nothing), ProtoMetaAttribs(2, :step, :int64, Int64, 0, false, [0], nothing), ProtoMetaAttribs(3, :file_version, :string, AbstractString, 0, false, [""], nothing), ProtoMetaAttribs(4, :graph_def, :bytes, Array{UInt8,1}, 0, false, Any[UInt8[]], nothing), ProtoMetaAttribs(5, :summary, :obj, TensorBoardLogger.tensorboard.Summary, 0, false, Any[], ProtoMeta for TensorBoardLogger.tensorboard.Summary
ProtoMetaAttribs[ProtoMetaAttribs(1, :value, :obj, Array{TensorBoardLogger.tensorboard.Summary_Value,1}, 2, false, Any[TensorBoardLogger.tensorboard.Summary_Value[]], ProtoMeta for TensorBoardLogger.tensorboard.Summary_Value
ProtoMetaAttribs[ProtoMetaAttribs(7, :node_name, :string, AbstractString, 0, false, [""], nothing), ProtoMetaAttribs(1, :tag, :string, AbstractString, 0, false, [""], nothing), ProtoMetaAttribs(9, :metadata, :obj, TensorBoardLogger.tensorboard.SummaryMetadata, 0, false, Any[], ProtoMeta for TensorBoardLogger.tensorboard.SummaryMetadata
ProtoMetaAttribs[ProtoMetaAttribs(1, :plugin_data, :obj, TensorBoardLogger.tensorboard.SummaryMetadata_PluginData, 0, false, Any[], ProtoMeta for TensorBoardLogger.tensorboard.SummaryMetadata_PluginData
ProtoMetaAttribs[ProtoMetaAttribs(1, :plugin_name, :string, AbstractString, 0, false, [""], nothing), ProtoMetaAttribs(2, :content, :bytes, Array{UInt8,1}, 0, false, Any[UInt8[]], nothing)]
), ProtoMetaAttribs(2, :display_name, :string, AbstractString, 0, false, [""], nothing), ProtoMetaAttribs(3, :summary_description, :string, AbstractString, 0, false, [""], nothing), ProtoMetaAttribs(4, :data_class, :int32, Int32, 0, false, Int32[0], nothing)]
), ProtoMetaAttribs(2, :simple_value, :float, Float32, 0, false, Float32[0.0], nothing), ProtoMetaAttribs(3, :obsolete_old_style_histogram, :bytes, Array{UInt8,1}, 0, false, Any[UInt8[]], nothing), ProtoMetaAttribs(4, :image, :obj, TensorBoardLogger.tensorboard.Summary_Image, 0, false, Any[], ProtoMeta for TensorBoardLogger.tensorboard.Summary_Image
ProtoMetaAttribs[ProtoMetaAttribs(1, :height, :int32, Int32, 0, false, Int32[0], nothing), ProtoMetaAttribs(2, :width, :int32, Int32, 0, false, Int32[0], nothing), ProtoMetaAttribs(3, :colorspace, :int32, Int32, 0, false, Int32[0], nothing), ProtoMetaAttribs(4, :encoded_image_string, :bytes, Array{UInt8,1}, 0, false, Any[UInt8[]], nothing)]
), ProtoMetaAttribs(5, :histo, :obj, TensorBoardLogger.tensorboard.HistogramProto, 0, false, Any[], ProtoMeta for TensorBoardLogger.tensorboard.HistogramProto
ProtoMetaAttribs[ProtoMetaAttribs(1, :min, :double, Float64, 0, false, [0.0], nothing), ProtoMetaAttribs(2, :max, :double, Float64, 0, false, [0.0], nothing), ProtoMetaAttribs(3, :num, :double, Float64, 0, false, [0.0], nothing), ProtoMetaAttribs(4, :sum, :double, Float64, 0, false, [0.0], nothing), ProtoMetaAttribs(5, :sum_squares, :double, Float64, 0, false, [0.0], nothing), ProtoMetaAttribs(6, :bucket_limit, :double, Array{Float64,1}, 2, true, Any[Float64[]], nothing), ProtoMetaAttribs(7, :bucket, :double, Array{Float64,1}, 2, true, Any[Float64[]], nothing)]
), ProtoMetaAttribs(6, :audio, :obj, TensorBoardLogger.tensorboard.Summary_Audio, 0, false, Any[], ProtoMeta for TensorBoardLogger.tensorboard.Summary_Audio
ProtoMetaAttribs[ProtoMetaAttribs(1, :sample_rate, :float, Float32, 0, false, Float32[0.0], nothing), ProtoMetaAttribs(2, :num_channels, :int64, Int64, 0, false, [0], nothing), ProtoMetaAttribs(3, :length_frames, :int64, Int64, 0, false, [0], nothing), ProtoMetaAttribs(4, :encoded_audio_string, :bytes, Array{UInt8,1}, 0, false, Any[UInt8[]], nothing), ProtoMetaAttribs(5, :content_type, :string, AbstractString, 0, false, [""], nothing)]
), ProtoMetaAttribs(8, :tensor, :obj, TensorBoardLogger.tensorboard.TensorProto, 0, false, Any[], ProtoMeta for TensorBoardLogger.tensorboard.TensorProto
ProtoMetaAttribs[ProtoMetaAttribs(1, :dtype, :int32, Int32, 0, false, Int32[0], nothing), ProtoMetaAttribs(2, :tensor_shape, :obj, TensorBoardLogger.tensorboard.TensorShapeProto, 0, false, Any[], ProtoMeta for TensorBoardLogger.tensorboard.TensorShapeProto
ProtoMetaAttribs[ProtoMetaAttribs(2, :dim, :obj, Array{TensorBoardLogger.tensorboard.TensorShapeProto_Dim,1}, 2, false, Any[TensorBoardLogger.tensorboard.TensorShapeProto_Dim[]], ProtoMeta for TensorBoardLogger.tensorboard.TensorShapeProto_Dim
ProtoMetaAttribs[ProtoMetaAttribs(1, :size, :int64, Int64, 0, false, [0], nothing), ProtoMetaAttribs(2, :name, :string, AbstractString, 0, false, [""], nothing)]
), ProtoMetaAttribs(3, :unknown_rank, :bool, Bool, 0, false, Bool[0], nothing)]
), ProtoMetaAttribs(3, :version_number, :int32, Int32, 0, false, Int32[0], nothing), ProtoMetaAttribs(4, :tensor_content, :bytes, Array{UInt8,1}, 0, false, Any[UInt8[]], nothing), ProtoMetaAttribs(13, :half_val, :int32, Array{Int32,1}, 2, true, Any[Int32[]], nothing), ProtoMetaAttribs(5, :float_val, :float, Array{Float32,1}, 2, true, Any[Float32[]], nothing), ProtoMetaAttribs(6, :double_val, :double, Array{Float64,1}, 2, true, Any[Float64[]], nothing), ProtoMetaAttribs(7, :int_val, :int32, Array{Int32,1}, 2, true, Any[Int32[]], nothing), ProtoMetaAttribs(8, :string_val, :bytes, Array{Array{UInt8,1},1}, 2, false, Any[Array{UInt8,1}[]], nothing), ProtoMetaAttribs(9, :scomplex_val, :float, Array{Float32,1}, 2, true, Any[Float32[]], nothing), ProtoMetaAttribs(10, :int64_val, :int64, Array{Int64,1}, 2, true, Any[Int64[]], nothing), ProtoMetaAttribs(11, :bool_val, :bool, Array{Bool,1}, 2, true, Any[Bool[]], nothing), ProtoMetaAttribs(12, :dcomplex_val, :double, Array{Float64,1}, 2, true, Any[Float64[]], nothing), ProtoMetaAttribs(14, :resource_handle_val, :obj, Array{TensorBoardLogger.tensorboard.ResourceHandleProto,1}, 2, false, Any[TensorBoardLogger.tensorboard.ResourceHandleProto[]], ProtoMeta for TensorBoardLogger.tensorboard.ResourceHandleProto
ProtoMetaAttribs[ProtoMetaAttribs(1, :device, :string, AbstractString, 0, false, [""], nothing), ProtoMetaAttribs(2, :container, :string, AbstractString, 0, false, [""], nothing), ProtoMetaAttribs(3, :name, :string, AbstractString, 0, false, [""], nothing), ProtoMetaAttribs(4, :hash_code, :uint64, UInt64, 0, false, UInt64[0x0000000000000000], nothing), ProtoMetaAttribs(5, :maybe_type_name, :string, AbstractString, 0, false, [""], nothing), ProtoMetaAttribs(6, :dtypes_and_shapes, :obj, Array{TensorBoardLogger.tensorboard.ResourceHandleProto_DtypeAndShape,1}, 2, false, Any[TensorBoardLogger.tensorboard.ResourceHandleProto_DtypeAndShape[]], ProtoMeta for TensorBoardLogger.tensorboard.ResourceHandleProto_DtypeAndShape
ProtoMetaAttribs[ProtoMetaAttribs(1, :dtype, :int32, Int32, 0, false, Int32[0], nothing), ProtoMetaAttribs(2, :shape, :obj, TensorBoardLogger.tensorboard.TensorShapeProto, 0, false, Any[], ProtoMeta for TensorBoardLogger.tensorboard.TensorShapeProto
ProtoMetaAttribs[ProtoMetaAttribs(2, :dim, :obj, Array{TensorBoardLogger.tensorboard.TensorShapeProto_Dim,1}, 2, false, Any[TensorBoardLogger.tensorboard.TensorShapeProto_Dim[]], ProtoMeta for TensorBoardLogger.tensorboard.TensorShapeProto_Dim
ProtoMetaAttribs[ProtoMetaAttribs(1, :size, :int64, Int64, 0, false, [0], nothing), ProtoMetaAttribs(2, :name, :string, AbstractString, 0, false, [""], nothing)]
), ProtoMetaAttribs(3, :unknown_rank, :bool, Bool, 0, false, Bool[0], nothing)]
)]
)]
), ProtoMetaAttribs(15, :variant_val, :obj, Array{TensorBoardLogger.tensorboard.VariantTensorDataProto,1}, 2, false, Any[TensorBoardLogger.tensorboard.VariantTensorDataProto[]], ProtoMeta for TensorBoardLogger.tensorboard.VariantTensorDataProto
ProtoMetaAttribs[ProtoMetaAttribs(1, :type_name, :string, AbstractString, 0, false, [""], nothing), ProtoMetaAttribs(2, :metadata, :bytes, Array{UInt8,1}, 0, false, Any[UInt8[]], nothing), ProtoMetaAttribs(3, :tensors, :obj, Array{TensorBoardLogger.tensorboard.TensorProto,1}, 2, false, Any[TensorBoardLogger.tensorboard.TensorProto[]], ProtoMeta for TensorBoardLogger.tensorboard.TensorProto
ProtoMetaAttribs[#= circular reference @-4 =#]
)]
), ProtoMetaAttribs(16, :uint32_val, :uint32, Array{UInt32,1}, 2, true, Any[UInt32[]], nothing), ProtoMetaAttribs(17, :uint64_val, :uint64, Array{UInt64,1}, 2, true, Any[UInt64[]], nothing)]
)]
)]
), ProtoMetaAttribs(6, :log_message, :obj, TensorBoardLogger.tensorboard.LogMessage, 0, false, Any[], ProtoMeta for TensorBoardLogger.tensorboard.LogMessage
ProtoMetaAttribs[ProtoMetaAttribs(1, :level, :int32, Int32, 0, false, Int32[0], nothing), ProtoMetaAttribs(2, :message, :string, AbstractString, 0, false, [""], nothing)]
), ProtoMetaAttribs(7, :session_log, :obj, TensorBoardLogger.tensorboard.SessionLog, 0, false, Any[], ProtoMeta for TensorBoardLogger.tensorboard.SessionLog
ProtoMetaAttribs[ProtoMetaAttribs(1, :status, :int32, Int32, 0, false, Int32[0], nothing), ProtoMetaAttribs(2, :checkpoint_path, :string, AbstractString, 0, false, [""], nothing), ProtoMetaAttribs(3, :msg, :string, AbstractString, 0, false, [""], nothing)]
), ProtoMetaAttribs(8, :tagged_run_metadata, :obj, TensorBoardLogger.tensorboard.TaggedRunMetadata, 0, false, Any[], ProtoMeta for TensorBoardLogger.tensorboard.TaggedRunMetadata
ProtoMetaAttribs[ProtoMetaAttribs(1, :tag, :string, AbstractString, 0, false, [""], nothing), ProtoMetaAttribs(2, :run_metadata, :bytes, Array{UInt8,1}, 0, false, Any[UInt8[]], nothing)]
), ProtoMetaAttribs(9, :meta_graph_def, :bytes, Array{UInt8,1}, 0, false, Any[UInt8[]], nothing)]
```

**Comment by PhilipVinc** [comment]
Event was generated from [this proto file](https://github.com/tensorflow/tensorboard/blob/6086f25a221a3ac40a84ddd9027dbe605ac2dcd9/tensorboard/compat/proto/event.proto#L15) , and it's Julia meta struct is [here](https://github.com/PhilipVinc/TensorBoardLogger.jl/blob/a372bc2c44cebd2dce464e17a5e34b0d26208fdb/src/protojl/tensorboard/event_pb.jl#L169), in case it might help shed some light.

**Comment by tanmaykm** [comment]
Thanks. I am able to replicate the issue. Will look into it.

**Comment by tanmaykm** [comment]
Should be fixed now with #150. Needs the protoc codegen to be run again though.

**Comment by PhilipVinc** [comment]
thanks a lot!
